mixer/adapter/circonus/circonus.go

```diff
@@ -0,0 +1,179 @@
+		metrics map[string]config.Params_MetricInfo_Type
+	}
+)
+
+// ensure types implement the requisite interfaces
+var _ metric.HandlerBuilder = &builder{}
+var _ metric.Handler = &handler{}
+
+// bridge stdlog to env.Logger()
+type logToEnvLogger struct {
+	env adapter.Env
+}
+
+// Build constructs a circonus-gometrics instance and sets up the handler
+func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler, error) {
+
+	bridge := &logToEnvLogger{env: env}
+
+	cmc := &cgm.Config{
+		CheckManager: checkmgr.Config{
+			Check: checkmgr.CheckConfig{
+				SubmissionURL: b.adpCfg.SubmissionUrl,
+			},
+		},
+		Log:      log.New(bridge, "", 0),
+		Debug:    true,
+		Interval: "0s",
+	}
+
+	cm, err := cgm.NewCirconusMetrics(cmc)
+	if err != nil {
+		err = env.Logger().Errorf("Could not create NewCirconusMetrics: %v", err)
+		return nil, err
+	}
+
+	metrics := make(map[string]config.Params_MetricInfo_Type)
+	ac := b.adpCfg
+	for _, adpMetric := range ac.Metrics {
+		metrics[adpMetric.Name] = adpMetric.Type
+	}
+
+	return &handler{cm: *cm, env: env, metrics: metrics}, nil
+}
+
+// SetAdapterConfig assigns operator configuration to the builder
+func (b *builder) SetAdapterConfig(cfg adapter.Config) {
+	b.adpCfg = cfg.(*config.Params)
+}
+
+// Validate checks the SubmissionUrl parameter is well formed, and needed metrics are declared
+func (b *builder) Validate() (ce *adapter.ConfigErrors) {
+
+	// verify the submission url is well formed
+	if _, err := url.ParseRequestURI(b.adpCfg.SubmissionUrl); err != nil {
+		ce = ce.Append("submission_url", err)
+	}
+
+	// put the metric config into a map for use below
+	configMetrics := make(map[string]config.Params_MetricInfo_Type)
+	for _, configMetric := range b.adpCfg.Metrics {
+		configMetrics[configMetric.Name] = configMetric.Type
+	}
+
+	// verify there are no metric types without a corresponding config item
+	for metricName := range b.metricTypes {
+		if _, ok := configMetrics[metricName]; !ok {
+			err := fmt.Errorf("missing metric configuration %v", metricName)
+			ce = ce.Append("metrics", err)
+		}
+	}
+
+	return
+}
+
+// SetMetricTypes sets the available metric types in the builder
+func (b *builder) SetMetricTypes(types map[string]*metric.Type) {
+	b.metricTypes = types
+}
+
+// HandleMetric submits metrics to Circonus via circonus-gometrics
+func (h *handler) HandleMetric(ctx context.Context, insts []*metric.Instance) error {
+
+	for _, inst := range insts {
+
+		metricName := inst.Name
+		metricType := h.metrics[metricName]
+
+		switch metricType {
+
+		case config.GAUGE:
+			value, _ := inst.Value.(int64)
+			h.cm.Gauge(metricName, value)
+
+		case config.COUNTER:
+			h.cm.Increment(metricName)
+
+		case config.DISTRIBUTION:
+			value, _ := inst.Value.(time.Duration)
+			h.cm.Timing(metricName, float64(value))
+		}
+
+	}
+
+	h.env.ScheduleWork(h.cm.Flush)
+
+	return nil
+}
+
+// Executes any teardown, none needed for this adapter
+func (h *handler) Close() error {
+	return nil
```

Comment:
You probably should synchronously flush here, otherwise your buffered data would never make it out.

Reply:
Thanks for the spot, will make that update.

Reply:
I just realized you're flushing for every HandleMetric call, so there is no buffered state to flush in Close...